File: KataFour/KataFour.py
import re
import sys


def smallestTempSpread(file_path='weather.dat', pattern=r'\s+(\d+)\D+(\d+)\D+(\d+)'):
    regex = re.compile(pattern)
    minimum = sys.maxsize
    result = None
    with open(file_path) as f:
        for line in f:
            match = regex.match(line)
            if match:
                spread = int(match.group(2)) - int(match.group(3))
                if spread < minimum:
                    minimum = spread
                    result = match.group(1)
    return result


def smallestScoreSpread(file_path='football.dat', pattern=r'\s+\d+\.\s+([a-zA-Z_]+)(?:\D+\d+){4}\D+(\d+)\D+(\d+)'):
    regex = re.compile(pattern)
    minimum = sys.maxsize
    result = None
    with open(file_path) as f:
        for line in f:
            match = regex.match(line)
            if match:
                spread = abs(int(match.group(2)) - int(match.group(3)))
                if spread < minimum:
                    minimum = spread
                    result = match.group(1)
    return result


def smallest(file_path, pattern):
    regex = re.compile(pattern)
    minimum = sys.maxsize
    result = None
    with open(file_path) as f:
        for line in f:
            match = regex.match(line)
            if match:
                spread = abs(int(match.group(2)) - int(match.group(3)))
                if spread < minimum:
                    minimum = spread
                    result = match.group(1)
    return result


def smallestTempRefactor():
    return smallest(file_path='weather.dat', pattern=r'\s+(\d+)\D+(\d+)\D+(\d+)')


def smallestScoreRefactor():
    return smallest(file_path='football.dat', pattern=r'\s+\d+\.\s+([a-zA-Z_]+)(?:\D+\d+){4}\D+(\d+)\D+(\d+)')

File: ruby/04/KataFour.rb
class KataFour

  def self.smallest_temp_spread(file_path='weather.dat', regex=/\s+(\d+)\D+(\d+)\D+(\d+)/)
    minimum = 9999999999
    result = nil
    File.foreach(file_path) do |line|
      if match = regex.match(line)
        one, two, three = match.captures
        spread = (two.to_i - three.to_i).abs
        if spread < minimum
          minimum = spread
          result = one
        end
      end
    end
    result
  end


  def self.smallest_score_spread(file_path='football.dat', regex=/\s+\d+\.\s+([a-zA-Z_]+)(?:\D+\d+){4}\D+(\d+)\D+(\d+)/)
    minimum = 9999999999
    result = nil
    File.foreach(file_path) do |line|
      if match = regex.match(line)
        one, two, three = match.captures
        spread = (two.to_i - three.to_i).abs
        if spread < minimum
          minimum = spread
          result = one
        end
      end
    end
    result
  end


  def self.smallest(file_path, regex)
    minimum = 9999999999
    result = nil
    File.foreach(file_path) do |line|
      if match = regex.match(line)
        one, two, three = match.captures
        spread = (two.to_i - three.to_i).abs
        if spread < minimum
          minimum = spread
          result = one
        end
      end
    end
    result
  end


  def self.smallest_temp_refactor
    smallest(file_path='weather.dat', pattern=/\s+(\d+)\D+(\d+)\D+(\d+)/)
  end

  def self.smallest_score_refactor
    smallest(file_path='football.dat', pattern=/\s+\d+\.\s+([a-zA-Z_]+)(?:\D+\d+){4}\D+(\d+)\D+(\d+)/)
  end
end